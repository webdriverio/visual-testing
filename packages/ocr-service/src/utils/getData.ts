import logger from '@wdio/logger'
import { getNodeOcrData, getSystemOcrData } from './tesseract.js'
import type { GetOcrData, Line, GetData, GetDataOptions, RectReturn, Words } from '../types.js'
import { adjustElementBbox, getBase64ScreenshotSize, isRectanglesObject } from './index.js'
import { drawHighlightedWords, processImage } from './imageProcessing.js'

const log = logger('@wdio/ocr-service:getData')

export default async function getData(browser: WebdriverIO.Browser, options: GetDataOptions): Promise<GetData> {
    const {
        contrast,
        cliFile,
        haystack,
        isTesseractAvailable,
        language,
        ocrImagesPath,
    } = options

    try {
        let dpr = 1
        let screenshot
        if (!cliFile) {
            const screenSize = await browser.getWindowSize()
            screenshot = await browser.takeScreenshot()
            const { width } = getBase64ScreenshotSize(screenshot)
            dpr = width / screenSize.width
        } else {
            screenshot = cliFile
        }
        const { filePath } = await processImage({ contrast, isAndroid: browser.isAndroid, isIOS: browser.isIOS, ocrImagesPath, screenshot })
        let elementRectangles
        let croppedFilePath

        if (haystack) {
            elementRectangles = isRectanglesObject(haystack) ? haystack as RectReturn : await browser.getElementRect((await haystack as WebdriverIO.Element).elementId)
            croppedFilePath = (await processImage({
                contrast,
                elementRectangles,
                isAndroid: cliFile ? false : browser.isAndroid,
                isIOS: cliFile ? false : browser.isIOS,
                ocrImagesPath,
                screenshot,
            })).filePath
        }

        // OCR the image
        let ocrData: GetOcrData
        const start = process.hrtime()

        if (isTesseractAvailable) {
            log.info('Using system installed version of Tesseract')
            ocrData = await getSystemOcrData({ filePath: croppedFilePath || filePath, language })
        } else {
            log.info('Using NodeJS version of Tesseract')
            ocrData = await getNodeOcrData({ filePath: croppedFilePath || filePath, language })
        }

        if (haystack && elementRectangles) {
            ocrData.lines.forEach((line:Line) => {
                line.bbox = adjustElementBbox(line.bbox, elementRectangles)
            })

            ocrData.words.forEach((word:Words) => {
                word.bbox = adjustElementBbox(word.bbox, elementRectangles)
            })
        }

        const diff = process.hrtime(start)
        const processTime = ((diff[0] * 1000000 + diff[1] / 1000) / 1000000).toFixed(3)

        log.info(`It took '${processTime}s' to process the image.`)
        log.info(`The following text was found through OCR:\n\n${ocrData.text.replace(/[\r\n]{2,}/g, '\n')}`)

        // Get all words and highlight them
        const highlights = ocrData.words.map((word) => word.bbox)
        await drawHighlightedWords({ filePath, highlights })

        log.info(`OCR Image with found text can be found here:\n\n${filePath}`)

        return {
            dpr,
            filePath,
            ...ocrData,
        }
    } catch (e) {
        throw new Error(String(e))
    }
}
