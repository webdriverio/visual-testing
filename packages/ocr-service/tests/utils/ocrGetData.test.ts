import { join } from 'node:path'
import logger from '@wdio/logger'
import { browser as wdioBrowser } from '@wdio/globals'
import { describe, it, expect, vi, beforeEach } from 'vitest'
import getData from '../../src/utils/getData.js'
import * as Tesseract from '../../src/utils/tesseract.js'
import * as ImageProcessing from '../../src/utils/imageProcessing.js'
import { adjustElementBbox, getBase64ScreenshotSize, isRectanglesObject } from '../../src/utils/index.js'

const browser = wdioBrowser
const log = logger('test')
vi.mock('@wdio/logger', () => import(join(process.cwd(), '__mocks__', '@wdio/logger')))

vi.mock('../../src/utils/tesseract.js')
vi.mock('../../src/utils/imageProcessing.js')
vi.mock('../../src/utils/index.js')
vi.mock('webdriver-image-comparison/dist/helpers/utils', () => ({
    getBase64ScreenshotSize: vi.fn(() => ({ width: 1200 })),
}))
vi.mock('../../src/utils/tesseract.js', () => ({
    getNodeOcrData: vi.fn().mockResolvedValue({
        text: 'Sample OCR text',
        lines: [],
        words: []
    }),
    getSystemOcrData: vi.fn().mockResolvedValue({
        text: 'Sample OCR text',
        lines: [{ text: 'Sample OCR text', bbox: { left: 1, right: 1, top: 1, bottom: 1 } }],
        words: [
            { text: 'Sample', bbox: { left: 1, right: 1, top: 1, bottom: 1 } },
            { text: 'OCR', bbox: { left: 2, right: 2, top: 2, bottom: 2 } },
            { text: 'text', bbox: { left: 3, right: 3, top: 3, bottom: 3 } },
        ]
    })
}))
vi.mock('../../src/utils/imageProcessing.js', () => ({
    processImage: vi.fn().mockResolvedValue({ filePath: '/fake/path/image.png' }),
    drawHighlightedWords: vi.fn().mockResolvedValue('')
}))
vi.mock('@wdio/globals', () => ({
    browser: {
        isAndroid: false,
        isIOS: false,
        getElementRect:vi.fn().mockResolvedValue({ x: 10, y: 20, width: 300, height: 400 }),
        getBase64ScreenshotSize:vi.fn().mockReturnValue({ width: 1200 }),
        getWindowSize: vi.fn().mockResolvedValue({ width: 1200, height: 800 }),
        takeScreenshot:vi.fn().mockResolvedValue('screenshotData'),
    }
}))

describe('getData', () => {
    beforeEach(() => {
        vi.clearAllMocks()
    })

    it('processes OCR data correctly using system Tesseract when available', async () => {
        const logInfoMock = vi.spyOn(log, 'info')
        const options = {
            contrast: 1.0,
            isTesseractAvailable: true,
            language: 'ENG',
            ocrImagesPath: '/fake/ocr/path'
        }
        vi.mocked(getBase64ScreenshotSize).mockReturnValue({ height: 1200, width: 1200 })

        const result = await getData(browser, options)

        expect(browser.getWindowSize).toHaveBeenCalled()
        expect(browser.takeScreenshot).toHaveBeenCalled()
        expect(getBase64ScreenshotSize).toHaveBeenCalledWith('screenshotData')
        expect(isRectanglesObject).not.toHaveBeenCalled()
        expect(ImageProcessing.processImage).toHaveBeenCalledTimes(1)
        expect(Tesseract.getSystemOcrData).toHaveBeenCalledWith({
            filePath: '/fake/path/image.png',
            language: 'ENG'
        })
        expect(adjustElementBbox).not.toHaveBeenCalled()
        expect(logInfoMock.mock.calls).toMatchSnapshot()
        expect(result).toMatchSnapshot()
    })

    it('processes OCR data correctly using NodeJS Tesseract', async () => {
        const logInfoMock = vi.spyOn(log, 'info')
        const options = {
            contrast: 1.0,
            isTesseractAvailable: false,
            language: 'ENG',
            ocrImagesPath: '/fake/ocr/path'
        }
        vi.mocked(getBase64ScreenshotSize).mockReturnValue({ height: 1200, width: 1200 })

        const result = await getData(browser, options)

        expect(browser.getWindowSize).toHaveBeenCalled()
        expect(browser.takeScreenshot).toHaveBeenCalled()
        expect(getBase64ScreenshotSize).toHaveBeenCalledWith('screenshotData')
        expect(isRectanglesObject).not.toHaveBeenCalled()
        expect(ImageProcessing.processImage).toHaveBeenCalledTimes(1)
        expect(Tesseract.getNodeOcrData).toHaveBeenCalledWith({
            filePath: '/fake/path/image.png',
            language: 'ENG'
        })
        expect(adjustElementBbox).not.toHaveBeenCalled()
        expect(logInfoMock.mock.calls).toMatchSnapshot()
        expect(result).toMatchSnapshot()
    })

    it('processes OCR data correctly using a haystack', async () => {
        const logInfoMock = vi.spyOn(log, 'info')
        const options = {
            contrast: 1.0,
            isTesseractAvailable: true,
            haystack: { x: 5, y: 5, width: 50, height: 50 },
            language: 'ENG',
            ocrImagesPath: '/fake/ocr/path'
        }
        vi.mocked(getBase64ScreenshotSize).mockReturnValue({ height: 1200, width: 1200 })

        const result = await getData(browser, options)
        const sanitizedCalls = logInfoMock.mock.calls.map((args) =>
            args.map((msg) =>
                typeof msg === 'string'
                    ? msg.replace(/'[\d.]+s'/g, "'X.XXXs'")
                    : msg
            )
        )

        expect(browser.getWindowSize).toHaveBeenCalled()
        expect(browser.takeScreenshot).toHaveBeenCalled()
        expect(getBase64ScreenshotSize).toHaveBeenCalledWith('screenshotData')
        expect(isRectanglesObject).toHaveBeenCalledTimes(1)
        expect(ImageProcessing.processImage).toHaveBeenCalledTimes(2)
        expect(Tesseract.getSystemOcrData).toHaveBeenCalledWith({
            filePath: '/fake/path/image.png',
            language: 'ENG'
        })
        expect(adjustElementBbox).toHaveBeenCalledTimes(4)
        expect(sanitizedCalls).toMatchSnapshot()
        expect(result).toMatchSnapshot()
    })

    it('should handle haystack that requires fetching RectReturn from browser.getElementRect', async () => {
        const logInfoMock = vi.spyOn(log, 'info')
        const options = {
            contrast: 1.0,
            isTesseractAvailable: true,
            language: 'ENG',
            ocrImagesPath: '/fake/ocr/path',
            haystack: { elementId: 'element123' } as WebdriverIO.Element
        }
        // @ts-ignore
        isRectanglesObject.mockReturnValue(false)
        browser.getElementRect = vi.fn().mockResolvedValue({ x: 10, y: 20, width: 300, height: 400 })
        const result = await getData(browser, options)

        expect(isRectanglesObject).toHaveBeenCalledWith(options.haystack)
        expect(browser.getElementRect).toHaveBeenCalledWith('element123')
        expect(result).toHaveProperty('filePath', '/fake/path/image.png')
        expect(logInfoMock.mock.calls).toMatchSnapshot()
    })

    it('handles errors when screenshot capture fails', async () => {
        const options = {
            contrast: 1.0,
            haystack: { x: 5, y: 5, width: 50, height: 50 },
            isTesseractAvailable: true,
            language: 'ENG',
            ocrImagesPath: '/fake/ocr/path'
        }
        browser.takeScreenshot = vi.fn().mockRejectedValue(new Error('Failed to take screenshot'))

        await expect(getData(browser, options)).rejects.toThrow('Failed to take screenshot')
    })
})
