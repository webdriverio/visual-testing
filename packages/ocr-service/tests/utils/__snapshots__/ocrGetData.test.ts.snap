// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getData > processes OCR data correctly using NodeJS Tesseract 1`] = `
[
  [
    "Using NodeJS version of Tesseract",
  ],
  [
    "It took '0.000s' to process the image.",
  ],
  [
    "The following text was found through OCR:

Sample OCR text",
  ],
  [
    "OCR Image with found text can be found here:

/fake/path/image.png",
  ],
]
`;

exports[`getData > processes OCR data correctly using NodeJS Tesseract 2`] = `
{
  "dpr": 1,
  "filePath": "/fake/path/image.png",
  "lines": [],
  "text": "Sample OCR text",
  "words": [],
}
`;

exports[`getData > processes OCR data correctly using a haystack 1`] = `
[
  [
    "Using system installed version of Tesseract",
  ],
  [
    "It took 'X.XXXs' to process the image.",
  ],
  [
    "The following text was found through OCR:

Sample OCR text",
  ],
  [
    "OCR Image with found text can be found here:

/fake/path/image.png",
  ],
]
`;

exports[`getData > processes OCR data correctly using a haystack 2`] = `
{
  "dpr": 1,
  "filePath": "/fake/path/image.png",
  "lines": [
    {
      "bbox": undefined,
      "text": "Sample OCR text",
    },
  ],
  "text": "Sample OCR text",
  "words": [
    {
      "bbox": undefined,
      "text": "Sample",
    },
    {
      "bbox": undefined,
      "text": "OCR",
    },
    {
      "bbox": undefined,
      "text": "text",
    },
  ],
}
`;

exports[`getData > processes OCR data correctly using system Tesseract when available 1`] = `
[
  [
    "Using system installed version of Tesseract",
  ],
  [
    "It took '0.000s' to process the image.",
  ],
  [
    "The following text was found through OCR:

Sample OCR text",
  ],
  [
    "OCR Image with found text can be found here:

/fake/path/image.png",
  ],
]
`;

exports[`getData > processes OCR data correctly using system Tesseract when available 2`] = `
{
  "dpr": 1,
  "filePath": "/fake/path/image.png",
  "lines": [
    {
      "bbox": {
        "bottom": 1,
        "left": 1,
        "right": 1,
        "top": 1,
      },
      "text": "Sample OCR text",
    },
  ],
  "text": "Sample OCR text",
  "words": [
    {
      "bbox": {
        "bottom": 1,
        "left": 1,
        "right": 1,
        "top": 1,
      },
      "text": "Sample",
    },
    {
      "bbox": {
        "bottom": 2,
        "left": 2,
        "right": 2,
        "top": 2,
      },
      "text": "OCR",
    },
    {
      "bbox": {
        "bottom": 3,
        "left": 3,
        "right": 3,
        "top": 3,
      },
      "text": "text",
    },
  ],
}
`;

exports[`getData > should handle haystack that requires fetching RectReturn from browser.getElementRect 1`] = `
[
  [
    "Using system installed version of Tesseract",
  ],
  [
    "It took '0.000s' to process the image.",
  ],
  [
    "The following text was found through OCR:

Sample OCR text",
  ],
  [
    "OCR Image with found text can be found here:

/fake/path/image.png",
  ],
]
`;
